File: cplus/leetcode/202509/algo3509.cpp
#include <bits/stdc++.h>
using namespace std;

// 回溯 - 4.7 搜索 
class Solution {
public:
    int maxProduct(vector<int>& nums, int k, int limit) {
        // 参考题解
        if (accumulate(nums.begin(), nums.end(), 0) < abs(k)) return -1;
        
        int ans = -1, n = nums.size(); 
        unordered_set<long long> vis;
        auto dfs = [&](this auto&& dfs, int i, int alt, int even, int prod, bool sel) -> void {
            long long mask = ((long long)i << 40) + (alt << 22)  + (even << 20) + (prod << 2) + sel;
            if (vis.count(mask)) return;
            vis.insert(mask);
            if (ans == limit or (ans>=0 and prod>limit)) return;
            if (i == n) {
                if (prod<=limit && sel && alt == k) ans = max(ans, prod);
                return;
            };

            dfs(i+1, alt, even, prod, sel);
            // min(prod*nums[i], limit+1) 压缩状态空间
            dfs(i+1, alt + (even==0?nums[i]:-nums[i]), even^1, min(prod*nums[i], limit+1), true);
        };
        dfs(0, 0, 0, 1, false); return ans;  
    }
};

File: python/algo/202509/algo3509.py
from functools import cache
from typing import List

# 回溯 - 4.7 搜索 
class Solution:
    def maxProduct(self, nums: List[int], k: int, limit: int) -> int:
        # 参考题解
        if (sum(nums) < k): return -1
        ans = -1
        n = len(nums)
        
        @cache
        def dfs(i: int, alt: int, even:int, prod:int, sel:bool):
            '''
            i: 当前的元素索引
            alt : 交错和
            even : 当前选择奇数或偶数个元素
            prod : 选择的元素的乘积
            sel : 是否包含已选择元素, 空元素不合法
            ''' 

            nonlocal ans
            if ans == limit or (ans >= 0 and prod > limit): # 剪枝
                return
            if i == n:
                if alt == k and prod <= limit and sel: ans = max(ans, prod)
                return
            
            # 不选择当前元素
            dfs(i + 1, alt, even, prod, sel)
            # 选择当前元素
            next_sum = alt + (nums[i] if even == 0 else -nums[i])
            # min(prod * nums[i], limit + 1) ！！！！最关键点，压缩状态空间 
            dfs(i + 1, next_sum, even ^ 1, min(prod * nums[i], limit + 1), True)
        
        dfs(0, 0, 0, 1, False)
        
        return ans